`BtcSwapTx`: update refund tx to consider all utxos

ok300 opened the pull request:
This PR updates `BtcSwapTx` and its `new_refund` method to consider all available utxos.

Fixes #72

ok300 committed:
BtcSwapTx: update refund tx to consider all utxos

ok300 commented:
@i5hi as I'm still testing this, I marked it as a draft and wanted to ask for your high level thoughts and feedback:
- is this the proper way of addressing this, or do you see a better way?
- should the two `BtcSwapTx` fields (`utxo` and `utxos`) be merged into one?
- are there any other implications or considerations for this PR that I might have missed?

i5hi commented:
The approach is good! 

We don't need the old `utxo` field anymore.

ok300 committed:
BtcSwapTx: remove utxo field, update struct methods

The BtcSwapTx methods are updated to use all available
inputs, where appropriate.

ok300 commented:
Alright, I continued with this approach and removed the `utxo` field in 3587e3c581f072282a6f3dbf88e2e3f4adfde3a1 .

Short summary of the changes:
- removed `utxo` field
- updated `new_refund` to use `fetch_utxos` instead of `fetch_utxo`
- updated `sign_claim` to rely only on the 1st utxo in `utxos` (*)
- updated `sign_refund`
  - cooperative branch: rely only on the 1st utxo in `utxos` (*)
  - non-cooperative branch: use all available utxos as inputs

I'm not fully sure about the parts marked with (*), more specifically:

* Can more than 1 utxos be claimed?
  * I assume no, because as soon as 1 utxo shows up on the lockup address, the Boltz Backend advances the swap to the next state.
* Can more than 1 utxos be refunded in cooperative mode?
  * I assume no, because this needs interaction with Boltz and intuitively it makes sense to only be able to refund what can be claimed (see above).

What do you think @i5hi? Are my assumptions correct?

i5hi commented:
- Can more than 1 utxo be claimed?
Technically, yes but in practice, No. 

Boltz will never lockup more than 1 utxo, so wont need to. However, for consistency the code should attempt to spend all utxos, incase for whatever reason there is more than 1 - no harm I suppose.

- Can more than 1 utxo be coop refunded?
Yes. 

We refund what we lock* up - so incase we lock up more than 1 utxo, we must refund both.

i5hi commented:
@michael1011 Need confirmation from you on this

i5hi commented:
Infact, we should stick with your original approach of using`fetch_utxo` for `new_claim` and always expect and spend just 1 utxo.

while `new_refund` will use `fetch_utxos` and always attempt to spend all.

michael1011 commented:
> Can more than 1 utxos be claimed?

In theory, yes. But it's very likely never going to happen (unless there is a grave bug) that we lock more than one UTXO for a swap.

> Can more than 1 utxos be refunded in cooperative mode?

Yes. When asking for a cooperative refund signature, one of the parameters is `index` which is the number of the input you want to have a partial signature for. That way, you can do multiple cooperative refunds (for a single swap or multiple ones) in a single transaction. When we have a swap in a failed state, we are happy to create as many refund signatures as you would like.

michael1011 commented on src/swaps/bitcoin.rs:
```diff
@@ -400,27 +400,36 @@ impl BtcSwapScript {
-        }
+    /// Fetch the first (utxo,amount) pair for the script_pubkey of this swap, which is typically
+    /// the lockup utxo. Returns None if no utxo for the script_pubkey is found.
+    pub fn fetch_utxo(
```
Should we get rid of this function completely?

ok300 replied:
Done: 82c6bfb526cba84cb6e1273dcb143b078b5be1ea

michael1011 commented on src/swaps/bitcoin.rs:
```diff
@@ -979,7 +1011,7 @@ impl BtcSwapTx {
+            let refund_tx_hex = serialize(&refund_tx_cooperative).to_lower_hex_string();
             let partial_sig_resp = match self.swap_script.swap_type {
                 SwapType::Chain => {
                     boltz_api.get_chain_partial_sig(&swap_id, &pub_nonce, &refund_tx_hex)
```
`get_chain_partial_sig` and `get_submarine_partial_sig` have `index` 0 hardcoded in them; we need to adjust those to have a parameter which needs to be the index of the input of the transaction

ok300 replied:
Done: e3f9c573b132eceef96465bbd811a1cbea0cfc61

michael1011 commented on src/swaps/bitcoin.rs:
```diff
@@ -1036,48 +1068,55 @@ impl BtcSwapTx {
 
             let mut witness = Witness::new();
             witness.push(final_schnorr_sig.to_vec());
+            refund_tx_cooperative.input[0].witness = witness;
```
We need a witness for *all* inputs we want to spend

ok300 replied:
Done: aff632f82e738b578eb618e7fce1ff91af78309d

One question I have here is this:

"Step 7: Get boltz's partial sig" starts by serializing the `refund_tx` to hex. It then calculates and attaches the witness for this input. This is inside a loop iterating over `refund_tx.input[input_index]`.

However this loop modifies the `refund_tx` at the end of each iteartion, by setting the witness for this input. So the tx hex will be different in each loop iteration. More specifically, it's this section: https://github.com/SatoshiPortal/boltz-rust/blob/e3f9c573b132eceef96465bbd811a1cbea0cfc61/src/swaps/bitcoin.rs#L995-L1060

I wonder, does the changing `refund_tx_hex` affect the partial sig calculations in Step 7, as they depend on this hex and are repeated for each input? Or does the partial sig logic account for this?

michael1011 replied:
It depends on what part of the transaction you are changing. Here we are just talking about the witness (= unlocking script here: https://wiki.bitcoinsv.io/index.php/SIGHASH_flags#Items_that_are_never_signed), which is not signed. You can ask for partial signatures of either the unsigned or partially signed transaction; both will work just fine

ok300 committed:
bitcoin.rs: Remove fetch_utxo in favor of fetch_utxos

ok300 committed:
BtcSwapTx new_refund: fetch utxo from Boltz only on error

This preserves the existing logic of when to fallback to querying Boltz for the utxo.

ok300 committed:
BtcSwapTx cooperative sign_refund: build and attach witness for all inputs

ok300 committed:
Add input_index arg for partial_sig methods

Add new argument for get_chain_partial_sig, get_submarine_partial_sig that represent the input index.

ok300 commented:
@i5hi @michael1011 thanks for your feedback. I addressed your points and added one question about `sign_refund`.

ok300 committed:
Fix regtest.rs tests

i5hi commented:
@ok300 Have you tried running the integration submarine tests and send 2 payments to the address to see if this all works well?

i5hi commented:
Also ensure that standard refunds are working as expected

ok300 commented:
@i5hi I ran `cargo test` and some tests passed, others were skipped with "ignored, requires invoice and refund address".

Are these the tests you mean and if so, what's the proper way to run them? Just change the invoice and refund address fields, then remove the `#ignore` annotation?

i5hi commented:
you dont need to remove ignore, for each submarine test you have to use a fresh invoice or you will get a 400 from boltz:

```
cargo test --test bitcoin bitcoin_v2_submarine -- --nocapture --ignored 
```

ok300 commented:
I tried `bitcoin_v2_submarine` with a fresh testnet invoice from Phoenix, but running the command you posted failed with "invoice not for current active network: testnet3".

Do I have to set anything else?

i5hi commented:
Just tried a pheonix testnet invoice; worked fine.

Try this

```
lntb5125180n1pnwmt4zpp5mvhqku7k4m4smvhdft67qq0kfpsg3jktsp99uaws60rq6m9wumfqcqpjsp5qlw0j3wlpnzx4cmrk3n8f09fqra2fmwjm4jk0j2l8uqeclag8c9q9q7sqqqqqqqqqqqqqqqqqqqsqqqqqysgqdqqmqz9gxqyjw5qrzjqwfn3p9278ttzzpe0e00uhyxhned3j5d9acqak5emwfpflp8z2cnflctr6qq3f9n3gqqqqlgqqqqqeqqjqeh4hcz22hnr9xc0tat79q0aujefdreewpseqvh6z4afk2qyd067r5z8zq863m5lgkj9nkrdmmtalf6j7zejla0xur3e22mq4dtev9tqpkhanae
```

ok300 committed:
bitcoin.rs sign_refund: fix prevout handling when refunding multiple utxos

ok300 committed:
bitcoin.rs sign_refund: fix input signing for non-coop refund

ok300 commented:
I ran two kinds of tests:

First, I used this branch to trigger (and complete) refunds for Incoming Chain Swaps (BTC -> LBTC) in https://github.com/breez/breez-sdk-liquid/pull/471 . Both cooperative and non-cooperative claims (to LBTC, as final step of the chain swap) and refunds (to BTC) were successful.

Second, I used @i5hi 's suggestion above to run the `bitcoin_v2_submarine` test. It essentially worked (the new invoice was paid), even though the test printed an error `HTTP("mempool min fee not met, 224 < 4543")` at L170, which is when calling `boltz_api_v2.post_submarine_claim_tx_details`. The testnet mempool fee estimation is generally erratic. I also didn't find a way to set custom (higher) feerates for the claim tx in the `bitcoin_v2_submarine` test. IMO this is a separate issue related to feerates on testnet.

The PR is ready for final review 🙏

ok300 committed:
Merge branch 'trunk' into ok300-chain-swap-refund-check-all-utoxs

ok300 committed:
Merge branch 'trunk' into ok300-chain-swap-refund-check-all-utoxs

ok300 committed:
Cargo fmt

ok300 commented:
@michael1011 @i5hi anything else I can do to help with the review?

We are using this branch in the Liquid Breez SDK and it would make life easier to get it merged 🙏

michael1011 reviewed (approved):
utACK

i5hi commented:
Thanks @ok300
